spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Container.scala

```diff
@@ -41,18 +42,92 @@ abstract class Container[A <: Activity : ClassTag,
     B <: Activity : ClassTag, T: ClassTag](
   implicit ev: TensorNumeric[T]) extends AbstractModule[A, B, T] {
 
+  private var parent: Container[A, B, T] = null
+
   // list of sub modules
   val modules: ArrayBuffer[AbstractModule[Activity, Activity, T]]
   = ArrayBuffer[AbstractModule[Activity, Activity, T]]()
 
+  /**
+   * We base on the order of this returning result for compiling.
+   * In general, it's a list of nodes from topology sort.
+   * return empty as do nothing by default.
+   */
+  private[bigdl] def compilingPath(): List[Node[AbstractModule[Activity, Activity, T]]] = {
+    val kmodels = modules.filter(_.isInstanceOf[KerasModule[A, B, T]])
+    if (kmodels.length > 0) {
+      throw new RuntimeException(
+        s"""Please do not use KerasModule: ${kmodels.mkString(",")}
+           | within Container other than Sequential and Graph""".stripMargin)
+    }
+    List()
+  }
+
+  private[bigdl] def toActivity(values: List[Activity]): Activity = {
+    if (values.isEmpty) {
+      return null
+    }
+    if (values.length == 1) {
+      return values(0)
+    } else {
+      val t = new Table()
+      values.foreach {v =>
+        t.insert(v)
+      }
+      return t
+    }
+  }
+
+  private[bigdl] def doCompile(executionNodes: List[ModuleNode[T]]): Unit = {
+    var i = 0
+    while (i < executionNodes.length) {
+      val node = executionNodes(i)
+      val preNodes = node.prevNodes
+      val inputShapes = if (preNodes.isEmpty) {
+        if (node.element.getInputShape() == null) {
+          throw new StartingInputException("The first layer should explicitly declare inputShape")
+        } else {
+          List(node.element.getInputShape())
+        }
+      } else {
+        preNodes.map{_.element.getOutputShape()}.toList
+      }
+      node.element.build(toActivity(inputShapes))
+      i += 1
+    }
+  }
+
+  private[bigdl] class StartingInputException(msg: String) extends RuntimeException(msg)
+
+  private[bigdl] final def compile(): Unit = {
+    val executionNodes = this.compilingPath()
+    try {
+      doCompile(executionNodes)
+      if (this.parent != null) {
+        this.parent.compile()
+      }
+    } catch {
+      case e: StartingInputException =>
+        // For pure old-style model, it's fine that it cann't be compiled for compatibility.
+        if (executionNodes.filter(_.element.isInstanceOf[KerasModule[A, B, T]]).length > 0) {
+          throw e
+        }
+      case e: Throwable => throw e
+    }
+  }
+
   /**
    * Add a sub-module to the contained `modules`
    *
    * @param module module to be add
    * @return this container
    */
   def add(module: AbstractModule[_ <: Activity, _ <: Activity, T]): this.type = {
+    if (module.isInstanceOf[Container[A, B, T]]) {
```

Comment:
module.isInstanceOf[Container[_, _, _]]

Reply:
What if user add this module into two different containers? should we check this and throw an exception then?

Reply:
what's the current logic for such behavior? we indeed need to throw exception if this layer cannot be shared.